packages/shared/src/server/ingestion/processEventBatch.ts

```diff
@@ -0,0 +1,400 @@
+import { randomUUID } from "crypto";
```

Comment:
@hassiebp is this copy-pasta? 🍝

Reply:
Yes! 

```bash
git diff main:web/src/pages/api/public/ingestion.ts packages/shared/src/server/ingestion/processEventBatch.ts
```

Here's the diff 

```diff
diff --git a/web/src/pages/api/public/ingestion.ts b/packages/shared/src/server/ingestion/processEventBatch.ts
index fb9d0064..183f3c97 100644
--- a/web/src/pages/api/public/ingestion.ts
+++ b/packages/shared/src/server/ingestion/processEventBatch.ts
@@ -1,54 +1,41 @@
-import { cors, runMiddleware } from "@/src/features/public-api/server/cors";
-import { type NextApiRequest, type NextApiResponse } from "next";
+import { randomUUID } from "crypto";
 import { z } from "zod";
-import { LangfuseNotFoundError } from "@langfuse/shared";
-import {
-  eventTypes,
-  ingestionEvent,
-  traceException,
-  redis,
-  logger,
-  handleBatch,
-  recordIncrement,
-  getCurrentSpan,
-  IngestionQueue,
-  LegacyIngestionQueue,
-  S3StorageService,
-  instrumentAsync,
-  getProcessorForEvent,
-  getClickhouseEntityType,
-  type AuthHeaderValidVerificationResult,
-  type LegacyIngestionEventType,
-  type IngestionEventType,
-} from "@langfuse/shared/src/server";
-import { telemetry } from "@/src/features/telemetry";
-import { jsonSchema } from "@langfuse/shared";
-import { isPrismaException } from "@/src/utils/exceptions";
-import { env } from "@/src/env.mjs";
+
 import {
   InvalidRequestError,
-  MethodNotAllowedError,
-  BaseError,
+  LangfuseNotFoundError,
   UnauthorizedError,
 } from "@langfuse/shared";
 import {
   addTracesToTraceUpsertQueue,
-  QueueJobs,
   instrumentSync,
+  QueueJobs,
 } from "@langfuse/shared/src/server";
-import { randomUUID } from "crypto";
-import { prisma } from "@langfuse/shared/src/db";
-import { tokenCount } from "@/src/features/ingest/usage";
-import { ApiAuthService } from "@/src/features/public-api/server/apiAuth";
-import { RateLimitService } from "@/src/features/public-api/server/RateLimitService";
 
-export const config = {
-  api: {
-    bodyParser: {
-      sizeLimit: "4.5mb",
-    },
-  },
-};
+import { env } from "../../env";
+import { AuthHeaderValidVerificationResult } from "../auth/types";
+import { getClickhouseEntityType } from "../clickhouse/schemaUtils";
+import {
+  getCurrentSpan,
+  instrumentAsync,
+  recordIncrement,
+  traceException,
+} from "../instrumentation";
+import { logger } from "../logger";
+import { LegacyIngestionEventType } from "../queues";
+import { IngestionQueue } from "../redis/ingestionQueue";
+import { LegacyIngestionQueue } from "../redis/legacyIngestion";
+import { redis } from "../redis/redis";
+import { S3StorageService } from "../services/S3StorageService";
+import { handleBatch } from "./legacy";
+import { getProcessorForEvent } from "./legacy/EventProcessor";
+import { eventTypes, ingestionEvent, IngestionEventType } from "./types";
+import { type Model } from "../../db";
+
+export type TokenCountDelegate = (p: {
+  model: Model;
+  text: unknown;
+}) => number | undefined;
 
 let s3StorageServiceClient: S3StorageService;
 
@@ -66,133 +53,10 @@ const getS3StorageServiceClient = (bucketName: string): S3StorageService => {
   return s3StorageServiceClient;
 };
 
-/**
- * This handler performs multiple actions to ingest data. It is compatible with the new async workflow, but also
- * supports the old synchronous workflow which processes events in the web container.
- * Overall, the processing of each incoming request happens in three stages
- * 1. Validation
- *   - Check that the user has permissions
- *   - Check whether rate-limits are breached
- *   - Check that the request is well-formed
- * 2. Async Processing
- *   - Upload each event to S3 for long-term storage and as an event cache
- *   - Add the event batch to the queue for async processing
- *   - Fallback to sync processing on errors
- * 3. Sync Processing
- * The last two stages live in a processEventBatch function which is reused for the POST scores endpoint and for
- * legacy event types.
- */
-export default async function handler(
-  req: NextApiRequest,
-  res: NextApiResponse,
-) {
-  try {
-    await runMiddleware(req, res, cors);
-
-    // add context of api call to the span
-    const currentSpan = getCurrentSpan();
-
-    // get x-langfuse-xxx headers and add them to the span
-    Object.keys(req.headers).forEach((header) => {
-      if (
-        header.toLowerCase().startsWith("x-langfuse") ||
-        header.toLowerCase().startsWith("x_langfuse")
-      ) {
-        currentSpan?.setAttributes({
-          [`langfuse.header.${header.slice(11).toLowerCase().replaceAll("_", "-")}`]:
-            req.headers[header],
-        });
-      }
-    });
-
-    if (req.method !== "POST") throw new MethodNotAllowedError();
-
-    // CHECK AUTH FOR ALL EVENTS
-    const authCheck = await new ApiAuthService(
-      prisma,
-      redis,
-    ).verifyAuthHeaderAndReturnScope(req.headers.authorization);
-
-    if (!authCheck.validKey) {
-      throw new UnauthorizedError(authCheck.error);
-    }
-
-    try {
-      const rateLimitCheck = await new RateLimitService(redis).rateLimitRequest(
-        authCheck.scope,
-        "ingestion",
-      );
-
-      if (rateLimitCheck?.isRateLimited()) {
-        return rateLimitCheck.sendRestResponseIfLimited(res);
-      }
-    } catch (e) {
-      // If rate-limiter returns an error, we log it and continue processing.
-      // This allows us to fail open instead of reject requests.
-      logger.error("Error while rate limiting", e);
-    }
-
-    const batchType = z.object({
-      batch: z.array(z.unknown()),
-      metadata: jsonSchema.nullish(),
-    });
-
-    const parsedSchema = instrumentSync(
-      { name: "ingestion-zod-parse-unknown-batch-event" },
-      () => batchType.safeParse(req.body),
-    );
-
-    if (!parsedSchema.success) {
-      logger.info("Invalid request data", parsedSchema.error);
-      return res.status(400).json({
-        message: "Invalid request data",
-        errors: parsedSchema.error.issues.map((issue) => issue.message),
-      });
-    }
-
-    await telemetry();
-
-    const result = await processEventBatch(parsedSchema.data.batch, authCheck);
-    return res.status(207).json(result);
-  } catch (error: unknown) {
-    if (!(error instanceof UnauthorizedError)) {
-      logger.error("error_handling_ingestion_event", error);
-      traceException(error);
-    }
-
-    if (error instanceof BaseError) {
-      return res.status(error.httpCode).json({
-        error: error.name,
-        message: error.message,
-      });
-    }
-
-    if (isPrismaException(error)) {
-      return res.status(500).json({
-        error: "Internal Server Error",
-      });
-    }
-
-    if (error instanceof z.ZodError) {
-      logger.error(`Zod exception`, error.errors);
-      return res.status(400).json({
-        message: "Invalid request data",
-        error: error.errors,
-      });
-    }
-
-    const errorMessage =
-      error instanceof Error ? error.message : "An unknown error occurred";
-    res.status(500).json({
-      message: "Invalid request data",
-      errors: [errorMessage],
-    });
-  }
-}
-
 export const processEventBatch = async (
   input: unknown[],
   authCheck: AuthHeaderValidVerificationResult,
+  tokenCountDelegate: TokenCountDelegate,
 ): Promise<{
   successes: { id: string; status: number }[];
   errors: {
@@ -237,7 +101,7 @@ export const processEventBatch = async (
         });
         return [];
       }
-      if (!isAuthorized(parsed.data, authCheck)) {
+      if (!isAuthorized(parsed.data, authCheck, tokenCountDelegate)) {
         authenticationErrors.push({
           id: parsed.data.id,
           error: new UnauthorizedError("Access Scope Denied"),
@@ -425,7 +289,7 @@ export const processEventBatch = async (
   /*******************
    * SYNC PROCESSING *
    *******************/
-  const result = await handleBatch(sortedBatch, authCheck, tokenCount);
+  const result = await handleBatch(sortedBatch, authCheck, tokenCountDelegate);
 
   await addTracesToTraceUpsertQueue(result.results, authCheck.scope.projectId);
 
@@ -439,9 +303,10 @@ export const processEventBatch = async (
 const isAuthorized = (
   event: IngestionEventType,
   authScope: AuthHeaderValidVerificationResult,
+  tokenCountDelegate: TokenCountDelegate,
 ): boolean => {
   try {
-    getProcessorForEvent(event, tokenCount).auth(authScope.scope);
+    getProcessorForEvent(event, tokenCountDelegate).auth(authScope.scope);
     return true;
   } catch (error) {
     return false;

```